### PR TITLE
Fix build output cycle detection

### LINKED ISSUE #7718: Regression: new-build / new-app input output cycle detection
```console
$ oc new-app -f test/extended/fixtures/incremental-auth-build.json 
--> Deploying template push-pull-builder for "test/extended/fixtures/incremental-auth-build.json"
error: output image of "demo/internal-image:latest" must be different than input, set a different tag with --to
```

Problems:
- The message is bad: `oc new-app` doesn't have (and shouldn't have) a `--to` flag, only `oc new-build`.
- The input and output images are not the same:

  ```json
          "sourceStrategy": {
            "from": {
              "kind": "DockerImage",
              "name": "centos/ruby-22-centos7:latest"
            }
          }
```
  ```json
        "output": {
          "to": {
            "kind": "ImageStreamTag",
            "name": "internal-image:latest"
          }
        },
```

- The behavior is off: `oc new-app -f template.json` should print a warning and proceed, just like `oc new-build` when it detects circular input/output with an explicit `--to` flag.

Introduced in #7321.

### PR DESCRIPTION
**Pull request by rhcarvalho**
Fixes #7718

### CODE REVIEW

**Comment by rhcarvalho** [comment]
[test][extended:core(incremental)][extended:core(build can have Dockerfile input)]

**Comment by rhcarvalho** [comment]
@bparees PTAL

Previous runs of the incremental extended test has failed with timeouts.
I've verified locally that the template from the extended test and the original issue report works fine now:

```console
[vagrant@localhost origin]$ oc new-app test/extended/fixtures/incremental-auth-build.json 
--> Deploying template push-pull-builder for "test/extended/fixtures/incremental-auth-build.json"
--> Creating resources with label app=initial-build ...
    imagestream "internal-image" created
    buildconfig "initial-build" created
    buildconfig "internal-build" created
--> Success
    Build config "initial-build" does not include any automatic triggers, use 'oc start-build' to start a build.
    Build config "internal-build" does not include any automatic triggers, use 'oc start-build' to start a build.
    Run 'oc status' to view your app.
```

Also fixed the extended test that Cesar ran into:

```console
[vagrant@localhost origin]$ oc new-build -D 'FROM openshift/origin-base'
--> Found Docker image 8bb96e3 (14 minutes old) from Docker Hub for "openshift/origin-base"

    * An image stream will be created as "origin-base:latest" that will track the source image
    * A Docker build using a predefined Dockerfile will be created
      * The resulting image will be pushed to image stream "origin-base:latest"
      * Every time "origin-base:latest" changes a new build will be triggered

error: output image of "demo/origin-base:latest" must be different than input, set a different tag with --to
[vagrant@localhost origin]$ oc new-build -D 'FROM openshift/origin-base' --to origin-base:output
--> Found Docker image 8bb96e3 (17 minutes old) from Docker Hub for "openshift/origin-base"

    * An image stream will be created as "origin-base:latest" that will track the source image
    * A Docker build using a predefined Dockerfile will be created
      * The resulting image will be pushed to image stream "origin-base:output"
      * Every time "origin-base:latest" changes a new build will be triggered

--> Creating resources with label build=origin-base ...
    imagestream "origin-base" created
    buildconfig "origin-base" created
--> Success
    Build configuration "origin-base" created and build triggered.
    Run 'oc logs -f bc/origin-base' to stream the build progress.
```

I'd like to get this in today fixing incorrect errors from `oc new-app`, and as a follow up next week I have started writing more tests around `checkCircularReferences` covering the cases we discussed in https://github.com/openshift/origin/pull/7321/files#r53043941 and more.

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -1097,17 +1097,20 @@ func (c *AppConfig) run(acceptors app.Acceptors) (*AppResult, error) {
...
 	}
 
-	err = c.checkCircularReferences(objects)
-	if err != nil {
-		if err, ok := err.(app.CircularOutputReferenceError); ok {
-			if len(c.To) == 0 {
-				// Output reference was generated, return error.
-				return nil, fmt.Errorf("%v, set a different tag with --to", err)
+	// Only check circular references for `oc new-build`.
+	if c.ExpectToBuild {
```
Comment (rhcarvalho) [inline pkg/generate/app/cmd/newapp.go]: @bparees FYI this is the main change, limiting the check for only when we're running `oc new-build`.
We could alternatively also check invocations of new-app, but then limited to printing a warning.

**Comment by bparees** [comment]
lgtm noting followup needed to fix the incorrect circular error messages reported by new-app.
[merge]

**Comment by rhcarvalho** [comment]
> lgtm noting followup needed to fix the incorrect circular error messages reported by new-app.

@bparees could you please clarify? With this patch we should not report any error/warning with `oc new-app` (https://github.com/openshift/origin/pull/7936#discussion_r55853107). The original intention was that this check applies only to `new-build`.

The follow up I have is testing all combinations of strategy/from, triggers and output reference. I'm opening the follow up now as WIP so it won't be forgotten.

**Comment by bparees** [comment]
@rhcarvalho i'm referring to the second bullet of 
https://github.com/openshift/origin/issues/7718

in which two images were reported as being the same when they clearly aren't.

I realize you've solved the problem by not doing any checking for new-app, but the fact that the code thought those two things were the same is still really bad and needs to be addressed because presumably new-build could also hit that condition.

Also I did think you were going to add the warning to new-app (which again necessitates fixing the bad comparison logic or else we'll get spurious warnings)

**Comment by rhcarvalho** [comment]
@bparees thanks. Indeed the bad comparisons should vanish with the follow up. We can also add a warning to new-app, will do in the follow up.

**Comment by bparees** [comment]
@rhcarvalho :+1:

**Pull request merged**